src/types.ts

```diff
@@ -145,6 +178,11 @@ export class IntersectionType extends ListOfTypes implements Type {
     accept<T>(visitor: TypeVisitor<T>): T {
         return visitor.visitIntersection(this)
     }
+
+    equal(other: Type): boolean {
+        return other instanceof IntersectionType && super.isPermutation(other)
```

Comment:
using isPermutation here has good intentions, but that won't achieve true equality by itself.
some cases come to mind:
type A = A1 & A2
type B = B1 & A
type BEQ = B1 & A1 & A2
B would not be equal to BEQ, but they really are. There are many more and much more complex cases to cover.

bottomline, I think true type equality is not what we want here. we could lower the expectations and say: equal is true if both types were constructed with equal parts in the same order.
So revert back to this simpler equality and remove added methods such as Arrays.isPermutation, keeping the code footprint smaller

Reply:
addressed in  7b9c05b

Reply:
restored isPermutation since in the end, we need it to correctly test typeParser